native/com_wolfssl_WolfSSL.c

```diff
@@ -1336,6 +1336,76 @@ JNIEXPORT jstring JNICALL Java_com_wolfssl_WolfSSL_getAvailableCipherSuitesIana
     return retString;
 }
 
+#ifdef WOLFSSLJNI_USE_NATIVE_CRYPTOCB
+/**
+ * Default native wolfSSL crypto callback function.
+ * This is called by default when wolfJSSE's WolfSSLProvider.registerDevId()
+ * is called, and is called by the native JNI API below.
+ *
+ * This function should be directly edited here to meet required
+ * functionality, or re-implemented and the registration point in the following
+ * function below should be edited:
+ *
+ * Java_com_wolfssl_WolfSSL_wc_1CryptoCb_1RegisterDevice()
+ */
+int DefaultNativeCryptoDevCb(int devId, wc_CryptoInfo* info, void* ctx)
+{
+    int ret = CRYPTOCB_UNAVAILABLE;
+    (void)devId;
+    (void)info;
+    (void)ctx;
+
+    /* Return CRYPTOCB_UNAVAILABLE to bypass HW and use SW. Edit function
+     * body here for your correct/expected functionality. */
+    return ret;
+}
+#endif /* WOLFSSLJNI_USE_NATIVE_CRYPTOCB */
+
+JNIEXPORT jint JNICALL Java_com_wolfssl_WolfSSL_wc_1CryptoCb_1RegisterDevice
+  (JNIEnv* jenv, jclass jcl, jint devId)
+{
+#ifdef WOLF_CRYPTO_CB
+    #ifdef WOLFSSLJNI_USE_NATIVE_CRYPTOCB
```

Comment:
How would one register the default DefaultNativeCryptoDevCb for use alongside ccbVaultIc_CryptoDevCb, or is that even a potential use case? Part of what's neat about the callbacks is hot swapping out which crypto backend is used.

Reply:
Currently, these are mutually exclusive because the devId is being set by the caller for the default case, but it being used to lookup which native callback function to register.  This could be changed to make the DefaultNativeCryptoDevCb get registered with a fixed dev id, like 0x12345678.  Dunno.  In C, you can BYOCb function and the user can select the devId.  In Java, we had to instead compile in a set of fixed id's within JNI and hope the calling Java app knows which one to use.  Might need to reconsider this to maybe add a "query" function so the caller can see which native bottom ends are compiled in and are available to be registered.  At no point was the thought we'd allow a java class handle the callbacks, so the cryptocb feature in JNI is not as "neat".

Reply:
I agree with Bill, for now `WOLFSSLJNI_USE_NATIVE_CRYPTOCB` / `DefaultNativeCryptoDevCb()` is mutually exclusive with calling another crypto callback implementation in `Java_com_wolfssl_WolfSSL_wc_1CryptoCb_1RegisterDevice()`.  `WOLFSSLJNI_USE_NATIVE_CRYPTOCB` should be a rarely used case.  I think we can leave this as-is for now and expand/adjust this in the future if new use cases come up.

I added a few more code comments around this area to make it clear our preference is to have new HW crypto implementations done inside native wolfCrypt proper.